// @paydock/client-sdk/lib/components/env.js
export const ENV = {
    SANDBOX: 'sandbox',
    SANDBOX_KOVENA: 'sandbox-kovena',
    SANDBOX_DEMO: 'sandbox-demo',
    SANDBOX_DEMO_KOVENA: 'sandbox-demo-kovena',
    PROD: 'production',
    STAGING: 'staging',
    STAGING_2: 'staging_2',
    STAGING_3: 'staging_3',
    STAGING_4: 'staging_4',
    STAGING_5: 'staging_5',
};
const KOVENA_ENV = [ENV.SANDBOX_KOVENA, ENV.SANDBOX_DEMO_KOVENA];
const REGEX_ALIAS = '^([a-zA-Z0-9](?:(?:[a-zA-Z0-9-.]*(?!-)\\.(?![-.]))*[a-zA-Z0-9]+)?)$';
export const DEFAULT_ENV = ENV.SANDBOX;
export const DEFAULT_ALIAS = {
    PAYDOCK: 'paydock.com',
    KOVENA: 'kovena.com',
};
export class Env {
    constructor(configs, defaultEnv = DEFAULT_ENV) {
        this.configs = configs;
        this.setEnv(defaultEnv);
    }
    setEnv(env, alias) {
        if (!this.isValidMode(this.configs, env))
            throw new Error('unknown env: ' + env);
        this.env = env;
        if (alias && !alias.match(REGEX_ALIAS))
            throw new Error('invalid: ' + alias);
        if (alias)
            this.alias = alias;
        else if (KOVENA_ENV.indexOf(this.env) !== -1)
            this.alias = DEFAULT_ALIAS.KOVENA;
        else
            this.alias = DEFAULT_ALIAS.PAYDOCK;
    }
    getEnv() {
        return this.env;
    }
    getConf() {
        for (let index in this.configs) {
            if (!this.configs.hasOwnProperty(index))
                continue;
            if (this.configs[index].env === this.getEnv())
                return {
                    url: (this.configs[index].url.indexOf('localhost') !== -1) ? this.configs[index].url : this.configs[index].url + this.alias,
                    env: this.configs[index].env,
                };
        }
        throw new Error('invalid env');
    }
    isValidMode(configs, env) {
        for (let index in configs) {
            if (!configs.hasOwnProperty(index))
                continue;
            if (configs[index].env === env)
                return true;
        }
        return false;
    }
}
