// @paydock/client-sdk/lib/components/flypay-iframe-event.js
import { IFrameEvent } from './iframe-event';
export var FLYPAY_EVENT;
(function (FLYPAY_EVENT) {
    FLYPAY_EVENT["AFTER_LOAD"] = "after_load";
    FLYPAY_EVENT["UNAVAILABLE"] = "unavailable";
    FLYPAY_EVENT["START_LOADING"] = "start_loading";
    FLYPAY_EVENT["END_LOADING"] = "end_loading";
    FLYPAY_EVENT["UPDATE"] = "update";
    FLYPAY_EVENT["PAYMENT_SUCCESSFUL"] = "payment_successful";
    FLYPAY_EVENT["PAYMENT_IN_REVIEW"] = "payment_in_review";
    FLYPAY_EVENT["PAYMENT_ERROR"] = "payment_error";
})(FLYPAY_EVENT || (FLYPAY_EVENT = {}));
export class FlypayIframeEvent extends IFrameEvent {
    on(eventName, widgetId, cb) {
        for (let event in FLYPAY_EVENT) {
            if (!FLYPAY_EVENT.hasOwnProperty(event))
                continue;
            if (eventName === FLYPAY_EVENT[event]) {
                this.listeners.push({
                    event: eventName,
                    listener: cb,
                    widget_id: widgetId,
                });
            }
        }
    }
}
