// @paydock/client-sdk/lib/components/loader.js
export const TEMPLATE = `
    <div class="cs-loader">
      <div class="cs-loader-inner">
        <label>	●</label>
        <label>	●</label>
        <label>	●</label>
        <label>	●</label>
        <label>	●</label>
        <label>	●</label>
      </div>
    </div>
`;
export const STYLE = `
    <style> 
        .cs-loader {
          position: absolute;
          top: 0;
          left: 0;
          height: 100%;
          width: 100%;
        }
        
        .cs-loader-inner {
          transform: translateY(-50%);
          top: 50%;
          position: absolute;
          width: calc(100% - 200px);
          color: #8e8d8c;
          padding: 0 100px;
          text-align: center;
        }
        
        
        .cs-loader-inner label {
          font-size: 20px;
          opacity: 0;
          display:inline-block;
        }
        
        @keyframes lol {
          0% {
            opacity: 0;
            transform: translateX(-300px);
          }
          33% {
            opacity: 1;
            transform: translateX(0px);
          }
          66% {
            opacity: 1;
            transform: translateX(0px);
          }
          100% {
            opacity: 0;
            transform: translateX(300px);
          }
        }
        
        @-webkit-keyframes lol {
          0% {
            opacity: 0;
            -webkit-transform: translateX(-300px);
          }
          33% {
            opacity: 1;
            -webkit-transform: translateX(0px);
          }
          66% {
            opacity: 1;
            -webkit-transform: translateX(0px);
          }
          100% {
            opacity: 0;
            -webkit-transform: translateX(300px);
            -moz-transform: translateX(300px);
          }
        }
        
        .cs-loader-inner label:nth-child(6) {
          -webkit-animation: lol 3s infinite ease-in-out;
          -moz-animation: lol 3s infinite ease-in-out;
          animation: lol 3s infinite ease-in-out;
        }
        
        .cs-loader-inner label:nth-child(5) {
          -webkit-animation: lol 3s 100ms infinite ease-in-out;
          -moz-animation: lol 3s 100ms infinite ease-in-out;
          animation: lol 3s 100ms infinite ease-in-out;
        }
        
        .cs-loader-inner label:nth-child(4) {
          -webkit-animation: lol 3s 200ms infinite ease-in-out;
          -moz-animation: lol 3s 200ms infinite ease-in-out;
          animation: lol 3s 200ms infinite ease-in-out;
        }
        
        .cs-loader-inner label:nth-child(3) {
          -webkit-animation: lol 3s 300ms infinite ease-in-out;
          -moz-animation: lol 3s 300ms infinite ease-in-out;
          animation: lol 3s 300ms infinite ease-in-out;
        }
        
        .cs-loader-inner label:nth-child(2) {
          -webkit-animation: lol 3s 400ms infinite ease-in-out;
          -moz-animation: lol 3s 400ms infinite ease-in-out;
          animation: lol 3s 400ms infinite ease-in-out;
        }
        
        .cs-loader-inner label:nth-child(1) {
          -webkit-animation: lol 3s 500ms infinite ease-in-out;
          -moz-animation: lol 3s 500ms infinite ease-in-out;
          animation: lol 3s 500ms infinite ease-in-out;
        }
    </style>
`;
export const EXTRA_TEMPLATE = `
    <div class="circ">
      <div class="load">A little patience ...</div>
      <div class="hands"></div>
      <div class="body"></div>
      <div class="head">
        <div class="eye"></div>
      </div>
    </div>
`;
export const EXTRA_STYLE = `
    <style>
        html{width: 100%;height: 100%;}
        body{margin: 0px;padding: 0px;background-color: #111;}
        
        .eye{
          width: 20px; height: 8px;
          background-color: #eee;
          border-radius:0px 0px 20px 20px;
          position: relative;
          top: 40px;
          left: 10px;
          box-shadow:  40px 0px 0px 0px #eee;              
        }
        
        .head{
          -webkit-backface-visibility: hidden;
          -moz-backface-visibility: hidden;
          backface-visibility: hidden;          
          position: relative;
          margin: -250px auto;
          width: 80px; height: 80px;
          background-color: #111;
          border-radius:50px;
          box-shadow: inset -4px 2px 0px 0px #eee;
           -webkit-animation:node 1.5s infinite alternate;
          -webkit-animation-timing-function:ease-out;
          -moz-animation:node 1.5s infinite alternate;
          -moz-animation-timing-function:ease-out;
          animation:node 1.5s infinite alternate;
          animation-timing-function:ease-out;
        }
        .body{ 
          position: relative;
          margin: 90px auto;
          width: 140px; height: 120px;
          background-color: #111;
          border-radius: 50px/25px ;
          box-shadow: inset -5px 2px 0px 0px #eee;
          -webkit-animation:node2 1.5s infinite alternate;
          -webkit-animation-timing-function:ease-out;  
          -moz-animation:node2 1.5s infinite alternate;
          -moz-animation-timing-function:ease-out;  
          animation:node2 1.5s infinite alternate;
          animation-timing-function:ease-out; 
        }
        
        @keyframes node {0%{ top:0px; }50%{ top:10px; }100% { top:0px;} }
        @keyframes node2 {0%{ top:-5px; }50%{ top:10px; }100% { top:-5px;}}
        @-moz-keyframes node {0%{ top:0px; }50%{ top:10px; }100% { top:0px;} }
        @-moz-keyframes node2 {0%{ top:-5px; }50%{ top:10px; }100% { top:-5px;}}
        @-webkit-keyframes node {0%{ top:0px; }50%{ top:10px; }100% { top:0px;} }
        @-webkit-keyframes node2 {0%{ top:-5px; }50%{ top:10px; }100% { top:-5px;}}
      
               
        .circ{
          -webkit-backface-visibility: hidden;
          -moz-backface-visibility: hidden;
          backface-visibility: hidden;
           margin: 60px auto;
          width: 180px; height: 180px;
          background-color: #111;
          border-radius: 0px 0px 50px 50px;
          position: relative;
          z-index: -1;  
          left: 0%;
          top: 20%;
          overflow: hidden;
        }
        
        .hands{
          -webkit-backface-visibility: hidden;
          -moz-backface-visibility: hidden;
          backface-visibility: hidden;
          margin-top: 140px;
          width: 120px;height: 120px;
          position: absolute;
          background-color: #111;
          border-radius:20px;
          box-shadow:-1px -4px 0px 0px #eee;
          transform:rotate(45deg);
          -webkit-transform:rotate(45deg);
          -mox-transform:rotate(45deg);
          top:75%;left: 16%;
          z-index: 1;
          -webkit-animation:node2 1.5s infinite alternate;
          -webkit-animation-timing-function:ease-out;
          -moz-animation:node2 1.5s infinite alternate;
          -moz-animation-timing-function:ease-out;
          animation:node2 1.5s infinite alternate;
          animation-timing-function:ease-out;
        }
        
        .load{  position: absolute;
          width: 100px; height: 20px;
           margin: -10px auto;
           -webkit-font-smoothing: antialiased;
          -moz-font-smoothing: antialiased;
          font-smoothing: antialiased;
          font-family: 'Julius Sans One', sans-serif;
          font-size:30px;
          font-weight:400;
          color:#eee;
          left: 10%;
          top: 5%;
        }
    </style>
`;
