// @paydock/client-sdk/lib/components/vault-display-iframe-event.js
import { IFrameEvent } from './iframe-event';
var VAULT_DISPLAY_EVENT;
(function (VAULT_DISPLAY_EVENT) {
    VAULT_DISPLAY_EVENT["AFTER_LOAD"] = "after_load";
    VAULT_DISPLAY_EVENT["SYSTEM_ERROR"] = "system_error";
    VAULT_DISPLAY_EVENT["CVV_SECURE_CODE_REQUESTED"] = "cvv_secure_code_requested";
    VAULT_DISPLAY_EVENT["CARD_NUMBER_SECURE_CODE_REQUESTED"] = "card_number_secure_code_requested";
    VAULT_DISPLAY_EVENT["ACCESS_FORBIDDEN"] = "access_forbidden";
    VAULT_DISPLAY_EVENT["SESSION_EXPIRED"] = "session_expired";
    VAULT_DISPLAY_EVENT["OPERATION_FORBIDDEN"] = "operation_forbidden";
})(VAULT_DISPLAY_EVENT || (VAULT_DISPLAY_EVENT = {}));
export class VaultDisplayIframeEvent extends IFrameEvent {
    on(eventName, widgetId, cb) {
        for (let event in VAULT_DISPLAY_EVENT) {
            if (!VAULT_DISPLAY_EVENT.hasOwnProperty(event))
                continue;
            if (eventName === VAULT_DISPLAY_EVENT[event]) {
                this.listeners.push({
                    event: eventName,
                    listener: cb,
                    widget_id: widgetId,
                });
            }
        }
    }
}
