// @paydock/client-sdk/lib/components/wallet-background.js
import { Background } from "./background";
const TEMPLATE = `
    <div class="checkout-container">
        <div class="checkout-bg-logo"></div>
        <a href="#" data-close>Close</a>
    </div>
`;
const STYLES = `
    .checkout-bg-logo {
        display: block;
        background: url({{url}}) no-repeat;
        width: 100px;
        height: 50px;
        margin: 0 auto;
        border-radius: 10px;
        background-size: contain;
    }
`;
export class WalletBackground extends Background {
    constructor(bgImageUrl) {
        super();
        this.bgImageUrl = bgImageUrl;
        this.imageStyle = null;
    }
    initControl() {
        if (!this.imageStyle)
            this.createImageStyles();
        super.initControl();
    }
    clear() {
        if (this.imageStyle)
            this.imageStyle.parentNode.removeChild(this.imageStyle);
        this.imageStyle = null;
        super.clear();
    }
    createTemplate() {
        let body = document.body || document.getElementsByTagName("body")[0];
        let template = String(TEMPLATE);
        this.overlay = document.createElement("div");
        this.overlay.classList.add("checkout-overlay");
        this.overlay.setAttribute("checkout-overlay", " ");
        this.overlay.innerHTML = template;
        body.appendChild(this.overlay);
        setTimeout(() => {
            if (this.isInit())
                this.overlay.classList.add("display");
        }, 5);
    }
    createImageStyles() {
        let head = document.head || document.getElementsByTagName("head")[0];
        let css = String(STYLES);
        let container = (document.querySelector(".checkout-container"));
        css = css.replace("{{url}}", this.bgImageUrl);
        this.imageStyle = document.createElement("style");
        this.imageStyle.type = "text/css";
        this.imageStyle.appendChild(document.createTextNode(css));
        head.appendChild(this.imageStyle);
    }
}
