// @paydock/client-sdk/lib/helper/browser.js
export class Browser {
    static isFacebook() {
        return (navigator.userAgent.indexOf('FBSN/iOS') !== -1
            && navigator.userAgent.indexOf('AppleWebKit') !== -1
            && navigator.userAgent.indexOf('(KHTML, like Gecko)') !== -1);
    }
    static isInstagram() {
        return (navigator.userAgent.indexOf('iOS') !== -1
            && navigator.userAgent.indexOf('Instagram') !== -1
            && navigator.userAgent.indexOf('(KHTML, like Gecko)') !== -1);
    }
    static isSupportPopUp() {
        return !this.isFacebook() && !this.isInstagram();
    }
    static getLanguage() {
        return window.navigator.language || '';
    }
    static getTimezoneOffset() {
        return new Date().getTimezoneOffset();
    }
    static getBrowserName() {
        const sUsrAg = navigator.userAgent;
        // "Mozilla/5.0 (X11; Ubuntu; Linux x86_64; rv:61.0) Gecko/20100101 Firefox/61.0"
        if (sUsrAg.indexOf('Firefox') > -1)
            return 'Mozilla Firefox';
        if (sUsrAg.indexOf('Opera') > -1)
            return 'Opera';
        // "Mozilla/5.0 (Windows NT 10.0; WOW64; Trident/7.0; .NET4.0C; .NET4.0E; Zoom 3.6.0; wbx 1.0.0; rv:11.0) like Gecko"
        if (sUsrAg.indexOf('Trident') > -1)
            return 'Microsoft Internet Explorer';
        // "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/58.0.3029.110 Safari/537.36 Edge/16.16299"
        if (sUsrAg.indexOf('Edge') > -1)
            return 'Microsoft Edge';
        // "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Ubuntu Chromium/66.0.3359.181 Chrome/66.0.3359.181 Safari/537.36"
        if (sUsrAg.indexOf('Chrome') > -1)
            return 'Google Chrome or Chromium';
        // "Mozilla/5.0 (iPhone; CPU iPhone OS 11_4 like Mac OS X) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/11.0 Mobile/15E148 Safari/604.1 980x1306"
        if (sUsrAg.indexOf('Safari') > -1)
            return 'Apple Safari';
        return 'unknown';
    }
    static isJavaEnabled() {
        return navigator.javaEnabled();
    }
    static getColorDepth() {
        return screen.colorDepth;
    }
    static getScreenHeight() {
        return screen.height;
    }
    static getScreenWidth() {
        return screen.width;
    }
}
