// @paydock/client-sdk/lib/checkout-button/runner.js
import { Background, TRIGGER as BACKGROUND_TRIGGER } from "../components/background";
import { ENV, Env } from '../components/env';
export const EVENT = {
    SUCCESS: 'success',
    DECLINED: 'declined',
    CLOSE: 'close',
    REFERRED: 'referred',
    ERROR: 'error',
};
export class Runner {
    constructor() {
        this.customRedirectUti = null;
        this.background = new Background();
        this.background.onTrigger(BACKGROUND_TRIGGER.FOCUS, () => this.continue());
        this.background.onTrigger(BACKGROUND_TRIGGER.CLOSE, () => this.stop());
        this.widgetEnv = new Env([
            { env: ENV.SANDBOX, url: 'https://widget-sandbox.' },
            { env: ENV.SANDBOX_KOVENA, url: 'https://widget-sandbox.' },
            { env: ENV.SANDBOX_DEMO, url: 'https://widget-sandbox-demo.' },
            { env: ENV.SANDBOX_DEMO_KOVENA, url: 'https://widget-sandbox-demo.' },
            { env: ENV.PROD, url: 'https://widget.' },
            { env: ENV.STAGING, url: 'https://widsta.' },
            { env: ENV.STAGING_2, url: 'https://widsta-2.' },
            { env: ENV.STAGING_3, url: 'https://widsta-3.' },
            { env: ENV.STAGING_4, url: 'https://widsta-4.' },
            { env: ENV.STAGING_5, url: 'https://widsta-5.' },
        ]);
    }
    continue() {
    }
    next(checkoutData) {
    }
    stop() {
    }
    error(error, code, callback) {
        callback(true);
    }
    setCustomRedirectUri(uri) {
        this.customRedirectUti = uri;
    }
    setEnv(env, alias) {
        this.widgetEnv.setEnv(env, alias);
    }
    setBackgroundTitle(text) {
        this.background.setBackdropTitle(text);
    }
    setBackgroundDescription(text) {
        this.background.setBackdropDescription(text);
    }
    turnOffBackdrop() {
        this.background.turnOffControl();
        this.background.turnOffLoader();
    }
}
